File: plugins/aid/proj.android/src/org/cocos2dx/plugin/AdsAid.java
/****************************************************************************
Copyright (c) 2012-2013 cocos2d-x.org

http://www.cocos2d-x.org

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
package org.cocos2dx.plugin;

import java.util.Hashtable;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.view.View;

import jp.live_aid.aid.AdController;

public class AdsAid implements InterfaceAds {

	private static final String LOG_TAG = "AdsAid";
	private static Activity mContext = null;
	private static boolean bDebug = false;
    private static String AppId = "";
    private static String AppIdCp = "";
    private static String AppIdInterstitial = "";
    private static AdController _targetController;
    private static AdController _aidAdController;
    private static AdController _aidAdControllerCp;
    private static AdController _aidAdControllerInterstitial;
    private static AdsAid mAdapter = null;
    
	protected static void LogD(String msg) {
		if (bDebug) {
			Log.d(LOG_TAG, msg);
		}
	}

	public AdsAid(Context context) {
		mContext = (Activity) context;
        mAdapter = this;
	}

	@Override
	public void setDebugMode(boolean debug) {
		bDebug = debug;
	}

	@Override
	public void configDeveloperInfo(Hashtable<String, String> devInfo) {
        LogD("AdsAid configDeveloperInfo!");
        AppId = devInfo.get("AidID");
        AppIdCp = devInfo.get("AidIDCp");
        AppIdInterstitial = devInfo.get("AidIDInterstitial");
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                initAid();
                initAidCp();
                initAidInterstitial();
            }
        });
	}
    
    public void initAid() {
        LogD("AdsAid initAid!");
        Activity activity = mContext;
        if (!AppId.isEmpty()) {
            LogD("AdsAid gen _aidAdController!");
            _aidAdController = new AdController(AppId, activity);
            _aidAdController.setCreativeStyle(AdController.CreativeStyle.PLAIN_TEXT);
            _aidAdController.startPreloading();
        }
    }
    
    public void initAidCp() {
        LogD("AdsAid initAidCp!");
        Activity activity = mContext;
        if (!AppIdCp.isEmpty()) {
            LogD("AdsAid gen _aidAdControllerCp!");
            _aidAdControllerCp = new AdController(AppIdCp, activity) {
                
                //ポップアップの閉じるボタン押した
                protected void dialogCloseButtonWasClicked(android.app.Dialog dialog, View view) {
                    LogD("AdsAid dialogCloseButtonWasClicked!");
                    super.dialogCloseButtonWasClicked(dialog, view);
                    AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_AdsReceived, "tap_btn_close");
                }
                //ポップアップの詳細をみるボタン押した
                protected void dialogDetailButtonWasClicked(android.app.Dialog dialog, View view) {
                    LogD("AdsAid dialogDetailButtonWasClicked!");
                    super.dialogDetailButtonWasClicked(dialog, view);
                    AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_AdsReceived, "tap_btn_detail");
                }
                //backボタンでのポップアップキャンセルを不可にします
                protected void dialogDidCreated(android.app.Dialog dialog) {
                    dialog.setCancelable(false);
                }
            };
            _aidAdControllerCp.setCreativeStyle(AdController.CreativeStyle.PLAIN_TEXT);
            _aidAdControllerCp.startPreloading();
        }
    }

    public void initAidInterstitial() {
        LogD("AdsAid initAid!");
        Activity activity = mContext;
        if (!AppIdInterstitial.isEmpty()) {
            LogD("AdsAid gen _aidAdController!");
            _aidAdControllerInterstitial = new AdController(AppIdInterstitial, activity);
            _aidAdControllerInterstitial.setCreativeStyle(AdController.CreativeStyle.POPUP_IMAGE);
            _aidAdControllerInterstitial.startPreloading();
        }
    }
    
	@Override
	public void showAds(Hashtable<String, String> info, int pos) {
        final String mode = info.get("mode");
		PluginWrapper.runOnMainThread(new Runnable() {
			@Override
			public void run() {
                LogD("AdsAid showAds!");
                AdController.DialogType instType = AdController.DialogType.ON_DEMAND;
                if (mode != null && mode.equals("cp")) {
                    LogD("AdsAid cpMode!");
                    _targetController = _aidAdControllerCp;
                } else if (mode != null && mode.equals("interstitial")) {
                    _targetController = _aidAdControllerInterstitial;
                } else if (mode != null && mode.equals("exit")) {
                    _targetController = _aidAdControllerInterstitial;
                    instType = AdController.DialogType.ON_EXIT;
                } else {
                    LogD("AdsAid not cpMode!");
                    _targetController = _aidAdController;
                }
                
                if (_targetController != null) {
                    if (_targetController.hasLoadedContent()) {
                        _targetController.showDialog(instType);
                    } else {
                        final AdController.DialogType dt = instType;
                        new Thread(new Runnable(){
                            public void run(){
                                try
                                {
                                    long started = System.currentTimeMillis();
                                    while (true) {
                                        // 広告がロードされていなければ、最大10秒待つ
                                        if (_targetController.hasLoadedContent()) break;
                                        if (started + 10000L < System.currentTimeMillis()) break;
                                        Thread.sleep(500L);
                                    }
                                }
                                catch(InterruptedException iex) {}
                                
                                PluginWrapper.runOnMainThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        //広告ダイアログを表示
                                        _targetController.showDialog(dt);
                                    }
                                });
                            }
                        }).start();
                    }
                }
			}
		});
	}

	@Override
	public void hideAds(Hashtable<String, String> info) {
        LogD("Aid not support hideAds!");
	}

    @Override
	public void spendPoints(int points) {
		LogD("Aid not support spend points!");
	}
    
    @Override
	public void queryPoints() {
		LogD("Aid not support query points!");
	}
    
	@Override
	public String getSDKVersion() {
		return "1.1.1";
	}
    
	@Override
	public String getPluginVersion() {
		return "0.0.1";
	}

  public void startPreloading() {
      PluginWrapper.runOnMainThread(new Runnable() {
          @Override
          public void run() {
              if (_aidAdController != null) _aidAdController.startPreloading();
              if (_aidAdControllerCp != null) _aidAdControllerCp.startPreloading();
              if (_aidAdControllerInterstitial != null) _aidAdControllerInterstitial.startPreloading();
          }
      });
  }

  public void stopPreloading() {
      PluginWrapper.runOnMainThread(new Runnable() {
          @Override
          public void run() {
              if (_aidAdController != null) _aidAdController.stopPreloading();
              if (_aidAdControllerCp != null) _aidAdControllerCp.stopPreloading();
              if (_aidAdControllerInterstitial != null) _aidAdControllerInterstitial.stopPreloading();
          }
      });
  }
}

File: plugins/aid/proj.ios/AdsAid.m
/****************************************************************************
 Copyright (c) 2013 cocos2d-x.org
 
 http://www.cocos2d-x.org
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#import "AdsAid.h"
#import "AdsWrapper.h"

#define OUTPUT_LOG(...)     if (self.debug) NSLog(__VA_ARGS__);

@implementation AdsAid
{
    NSString* AppId;
    NSString* AppIdCp;
    NSString* AppIdInterstitial;
}

@synthesize debug = __debug;

#pragma mark InterfaceAds impl

//コールバック
- (void)aidCallback:(NSString*)msg code:(int)code;
{
    OUTPUT_LOG(@"AdsAid:aidCallback! code=%d", code);
    [AdsWrapper onAdsResult:self withRet:code withMsg:msg];
}

- (void)showDialogAfterLoad: (NSTimer*)timer
{
    NSDictionary* userInfo = [timer userInfo];
    AidAdAgent* agent = [userInfo objectForKey:@"agent"];
    
    do {
        if ([agent hasLoadedContent]) {
            // 広告コンテンツは読み込み済み
            OUTPUT_LOG(@"AidAd basic content is ready");
            break;
        }
        
        NSDate* started = [userInfo objectForKey:@"started"];
        NSTimeInterval delay = 0 - [started timeIntervalSinceNow];
        if ( delay > 10) {
            // 10秒経っていたら広告が間に合わないとしてあきらめる
            OUTPUT_LOG(@"AidAd content is not available");
            break;
        }
        
        // 次のタイマーによる呼び出しを待つ
        return;
        
    } while (0);
    
    //広告ダイアログを表示
    OUTPUT_LOG(@"AdsAid aidAgent showDialog");
    [agent showDialog];
    [timer invalidate];
}

- (void) configDeveloperInfo: (NSMutableDictionary*) devInfo
{
    OUTPUT_LOG(@"AdsAid configDeveloperInfo!");
    AppId = (NSString*) [devInfo objectForKey:@"AidID"];
    AppIdCp = (NSString*) [devInfo objectForKey:@"AidIDCp"];
    AppIdInterstitial = (NSString*) [devInfo objectForKey:@"AidIDInterstitial"];
    
    if (![AppId isEqualToString:@""]) {
        //広告の取得を開始します
        OUTPUT_LOG(@"AdsAid gen aidAgent");
        [[AidAd agentForMedia:AppId] startLoading];
    }
    if (![AppIdCp isEqualToString:@""]) {
        //自社広告の取得を開始します
        OUTPUT_LOG(@"AdsAid gen aidAgentCp");
        [[AidAd agentForMedia:AppIdCp] startLoading];
        [[AidAd agentForMedia:AppIdCp] setDelegate:self];
    }
    if (![AppIdInterstitial isEqualToString:@""]) {
        OUTPUT_LOG(@"AdsAid gen aidAgent");
        [[AidAd agentForMedia:AppIdInterstitial] setPreferredCreativeStyle:kAidAdCreativeStyle_POPUP_IMAGE];
        [[AidAd agentForMedia:AppIdInterstitial] startLoading];
    }
}

- (void) showAds: (NSMutableDictionary*) info position:(int) pos
{
    OUTPUT_LOG(@"AdsAid showAds!");
    NSString* mode = (NSString*)[info objectForKey:@"mode"];
    
    AidAdAgent* agent;
    if ([mode isEqualToString:@"cp"]) {
        agent = [AidAd agentForMedia:AppIdCp];
    } else if ([mode isEqualToString:@"interstitial"]) {
        agent = [AidAd agentForMedia:AppIdInterstitial];
    } else {
        agent = [AidAd agentForMedia:AppId];
    }
    
    if (agent)
    {
        if ([agent hasLoadedContent]) {
            [agent showDialog];
        } else {
            // 取得完了した事を定期的にチェックさせるタイマー
            NSDictionary* userInfo = @{@"agent":agent, @"started":[NSDate date]};
            [NSTimer scheduledTimerWithTimeInterval:1
                                             target:self
                                           selector:@selector(showDialogAfterLoad:)
                                           userInfo:userInfo
                                            repeats:YES];
        }
    }
}

- (void) hideAds: (NSMutableDictionary*) info
{
    OUTPUT_LOG(@"Aid not support hideAds!");
}

- (void) queryPoints
{
    OUTPUT_LOG(@"Aid not support query points!");
}

- (void) spendPoints: (int) points
{
    OUTPUT_LOG(@"Aid not support spend points!");
}

- (void) setDebugMode: (NSNumber*) isDebugMode
{
    self.debug = [isDebugMode boolValue];
}

- (NSString*) getSDKVersion
{
    return @"1.2.1";
}

- (NSString*) getPluginVersion
{
    return @"0.0.1";
}

- (void)adAgentDidOpenDialog:(AidAdAgent*)agent
{
    OUTPUT_LOG(@"AdsAid adAgentDidOpenDialog!");
}

- (void)adAgentDidCloseDialog:(AidAdAgent*)agent
{
    OUTPUT_LOG(@"AdsAid adAgentDidCloseDialog!");
}

- (void)adAgentDidDetectCloseButtonWasTapped:(AidAdAgent*)agent
{
    OUTPUT_LOG(@"AdsAid adAgentDidDetectCloseButtonWasTapped!");
    [self aidCallback:@"tap_btn_close" code:kAdsReceived];
}

- (void)adAgentDidDetectDetailButtonWasTapped:(AidAdAgent*)agent
{
    OUTPUT_LOG(@"AdsAid adAgentDidDetectDetailButtonWasTapped!");
    [self aidCallback:@"tap_btn_detail" code:kAdsReceived];
}

@end
